### PR TITLE
Improvements to `QuantumFloat`

### PR DESCRIPTION
**Pull request by PietropaoloFrisoni**
## Description

Cleans up `QuantumFloat`: adds type hints, fixes a silent exponent-type bug, replaces an exponential-time rounding path with an O(1) one, tightens exception types, and substantially expands test coverage.

## Related Issues

Closes #
Related to #

## Type of Change

- [ ] Feature (new functionality)
- [ ] Change Request (modification of existing functionality)
- [x] Bug Fix
- [ ] Refactoring (no behavior change)
- [x] Performance improvement
- [x] Documentation
- [ ] CI / Build

## Breaking Change?
- [ ] Yes
- [x] No

`QuantumFloat` methods now raise specific exception types (`TypeError`, `ValueError`, `NotImplementedError`) instead of a generic `Exception`. Code catching `except Exception:` is unaffected; code catching the generic `Exception` type specifically would need to broaden its `except` clause.

## What was changed?

- Added type hints across `QuantumFloat` (`src/qrisp/qtypes/quantum_float.py`) and `q_mult`/`hybrid_mult` (`src/qrisp/alg_primitives/arithmetic/SBP_arithmetic.py`), using `TYPE_CHECKING`-guarded imports to avoid circular imports.
- Fixed a bug where `QuantumFloat` addition/subtraction with different exponents could silently produce a `jax.Array` exponent instead of a plain `int` outside tracing, crashing later `2**exponent` calls.
- Replaced `encode(rounding=True)`'s exponential-time (`O(2^size)`) enumeration with a call to `truncate()`, an O(1) round-and-clip against the uniform representable-value grid (per reviewer feedback on this PR). Verified equivalent to the original across 13,000+ probe cases; the only divergence is round-half-to-even tie-breaking at exact midpoints instead of an arbitrary encoding-order artifact.
- Sped up `significant()` and `init_from()`.
- Replaced 26 generic `raise Exception(...)` call sites in `QuantumFloat` with specific exception types (`TypeError`, `ValueError`, `NotImplementedError`).
- Fixed `q_mult` silently returning `None` for an unrecognized `method` argument; now raises a `ValueError` naming the valid options. Gave `q_mult` a full docstring (it had none).
- Reverted an internal refactor (extracting the six comparison dunders' shared logic into a `_compare()` helper) after discovering it silently broke `with qf == 0:`-style conditional-environment usage in CI, by adding a stack frame that `adaptive_condition`'s fixed-depth caller-line lookup depends on. Added a regression test (`test_comparison_as_conditional_environment`) covering this usage pattern.
- Fixed stale/broken docstring examples, typos, and grammar throughout `QuantumFloat`'s documentation (e.g. a broken `:meth:` cross-reference from `exp_shift`'s docstring to `quantum_bit_shift`, "amount" → "number" for qubit counts, "inplace" → "in-place" for consistency, missing/misplaced punctuation). Verified via a full Sphinx HTML rebuild that all edited text renders correctly with no new warnings.
- Added 24 new tests to `tests/core_tests/test_quantum_float.py` (covering `__truediv__`, `__floordiv__`, `__pow__`'s inversion branch, in-place arithmetic, classical-scalar branches, `msize`, `sb_poly()`, `jdecoder()`, `quantum_bit_shift()`, the module-level `create_output_qf`/polynomial-encoding path, and the comparison-as-conditional-environment regression) and 2 new tests to `tests/jax_tests/test_jasp_QuantumFloat.py` (comparisons and arithmetic operators under Jasp tracing, previously untested).
- Fixed a real name-shadowing bug and a function-redeclaration bug found while adding tests (local `x`/`y` sympy symbols shadowing imported gate functions; two test functions both named `main`).
- Trimmed excessive multi-line class/method docstrings in the test files to single lines.
- Consolidated the changelog down to 3 entries for this PR (Improvements, Bug Fixes, Compatibility).

## How was it tested?

| Test-ID | Status |
|---------|--------|
| `tests/core_tests/test_quantum_float.py` | ✅ |
| `tests/jax_tests/test_jasp_QuantumFloat.py` | ✅ |
| `test_conditional_environments_example`, `test_custom_control` (CI regression, root-caused and fixed) | ✅ |
| `pyright`, `ruff format --check`, `ruff check`, `pylint` on all changed files | ✅ (no new findings) |
| `python -m doctest` on `quantum_float.py` | ✅ (all 105 doctest examples content-correct) |
| Sphinx HTML doc build | ✅ (no new warnings; edited text spot-checked in rendered output) |

## Checklist
- [x] My code follows the project's coding standards
- [x] I have performed a self-review
- [x] I have added/updated tests (referencing issue Test-IDs)
- [x] All tests pass locally and in CI
- [x] I have updated the documentation
- [x] I have added a changelog entry to `changelog-dev.rst`
- [x] Breaking changes are documented with migration path

## Reviewer Notes

### CODE REVIEW

**Comment by renezander90** [comment]
If I'm not mistaken, this PR also fixes #840 ?

There was a dedicated PR for this: #841 

However, this more comprehensive overhaul of the `QuantumFloat` is very welcome! Can #841 then be closed?
I would still prefer splitting the `create_output_qf` (sympy/float/modulus as done in #841).

File: src/qrisp/qtypes/quantum_float.py [hunk 1]
```diff
@@ -455,54 +548,100 @@ def sb_poly(self, m=0):
...
+            qubits. The default is False.
 
+        Returns
+        -------
+        None
+
+        """
+        value = encoding_number
+        if rounding:
+            # Round value to closest fitting number. Vectorized equivalent of
```
Comment (renezander90) [inline src/qrisp/qtypes/quantum_float.py]: I think this exponentially expensive logic is not necessary. See `truncate` method.

**Comment by PietropaoloFrisoni** [comment]
Thanks @renezander90 , I created this PR because I was studying `QuantumFloat` and I realized that I do not understand some of its components. I took the chance to open this PR in the meantime (there are many other improvements I am working on, always with the help of Claude). However this is not ready for review as there are many things I still need to check. If you want to go ahead with other PRs instead, feel free to go (I will resolve merge conflicts while they arise)

**Comment by purva-thakre** [comment]
@PietropaoloFrisoni Just letting you know I am trying to understand the `QuantumModulus` class. I will likely have a PR soon because the methods of this class lack proper docstrings. Wanted to make sure we don't double our efforts by working on the same module.

**Comment by PietropaoloFrisoni** [comment]
@purva-thakre Thanks for letting me know. I am focusing exclusively on `QuantumFloat` here (the changes in other files are just to suppress `pylance` warnings appearing here), so there should not be much overlap.

P.S. I am doing some refactorings to that module in [this PR](https://github.com/eclipse-qrisp/Qrisp/pull/827) as well

**Comment by renezander90** [comment]
@purva-thakre This PR: #827 refactoring parts of the infrastructure surrounding `QuantumModulus` could still be relevant. Regarding, `QuantumModulus`, some confusion within the class could arise from the fact that it's using two different modular arithmetic implementations: `jasp_arithmetic/jasp_montgomery` (new, Jasp compatible) and `modular_arithmetic/modular_multiplcation`(legacy, not Jasp compatible). We should further consolidate this. (I just discussed this yesterday with GPT.) This PR: #710 will eliminate some discrepancy in the Montgomery shift between the static/dynamic pipelines. Also tagging @diehoq.

Bildschirmfoto 2026-09-02 um 14 15 36

This is of course unrelated to the present PR, thus I apologise for spamming here!

P.S.: We should use issues or draft PR's or discord to keep aligned on which parts of the code base different people are working on to avoid duplicate work.

**Comment by renezander90** [comment]
Please also consider the issues that @positr0nium 's Claude pointed out on my PR: 
[repro_pr841.py](https://github.com/user-attachments/files/31747366/repro_pr841.py)

File: tests/jax_tests/test_jasp_QuantumFloat.py [hunk 2]
```diff
@@ -60,13 +60,104 @@ def inner(qf):
...
+    check_signed_static_exponent_dynamic()
+
+
+def test_jasp_QuantumFloat_comparisons():
+    """Test QuantumFloat's comparison operators under Jasp tracing."""
+    from qrisp import QuantumFloat
+    from qrisp.jasp import terminal_sampling
+
+    @terminal_sampling
+    def compare(a_value, b_value, op):
```
[the remark this answers is not in the document]
Comment (PietropaoloFrisoni) [reply tests/jax_tests/test_jasp_QuantumFloat.py:83]: This is false (checked locally). Resolving this

**Comment by PietropaoloFrisoni** [comment]
@renezander90 @purva-thakre  I will not request Copilot's reviews anymore since it's basically a rabbit hole. It keeps finding issues, and most of them are pre-existing edge cases that will never come up in practical usage, or they are out of scope for this PR (which was meant to be a quick refactoring but it already costed too much time and tokens)

File: src/qrisp/qtypes/quantum_float.py [hunk 3]
```diff
@@ -86,45 +97,44 @@ def _signed_int_iso_inv(y, n):
...
 
-# def signed_int_iso(x, n):
-#    if int(x) < -(2**n) or int(x) >= 2**n:
-#        raise Exception("Applying signed integer isomorphism resulted in overflow")
-
-#    if x >= 0:
-#        return x % 2**n
-#    else:
-#        return -abs(x) % 2 ** (n + 1)
+def trunc_poly(poly: sp.Expr, trunc_bounds: tuple[int, int]) -> sp.Expr:
```
Comment (renezander90) [inline src/qrisp/qtypes/quantum_float.py]: We should mark this as internal helper `_trunc_poly`

File: src/qrisp/qtypes/quantum_float.py [hunk 4]
```diff
@@ -355,43 +395,103 @@ def __init__(self, msize, exponent=0, qs=None, name=None, signed=False):
...
-        # represent.
+    def mshape(self) -> tuple[int | Array, int | Array]:
+        """The (log2(min), log2(max)) bounds of the absolute values this QuantumFloat can represent.
+
+        Returns
+        -------
+        tuple[int, int]
+            The (minimal, maximal) exponent of the representable magnitude.
+
+        """
```
Comment (renezander90) [inline src/qrisp/qtypes/quantum_float.py]:
```suggestion 
       """Return the mantissa's significance bounds.

        For a ``QuantumFloat`` with exponent ``k`` and mantissa size ``n``,
        the result is ``(k, k + n)``. The lower bound is inclusive and the
        upper bound is exclusive: mantissa qubit ``i`` has significance ``k + i``.

        Returns
        -------
        tuple[int | jax.Array, int | jax.Array]
            The inclusive lower and exclusive upper significance bounds.

        """
```

File: src/qrisp/qtypes/quantum_float.py [hunk 5]
```diff
@@ -355,43 +395,103 @@ def __init__(self, msize, exponent=0, qs=None, name=None, signed=False):
...
         Also validates that the input value can be represented within the bounds of the provided
         QuantumFloat in static mode.
+
+        .. note::
+
+            Unlike the base :meth:`QuantumVariable.encoder
+            <qrisp.QuantumVariable.encoder>`, this parameter is named ``i``
+            (not ``value``) for historical reasons specific to QuantumFloat.
+
+            Unlike the base's ``ScalarLike``, this does not accept ``complex``:
```
Comment (renezander90) [inline src/qrisp/qtypes/quantum_float.py]: I think the `base` is not typed at all currently? And it could also accept `str`(see `QuantumChar`). I would remove this part of the note. It should be cleat that one cannot instantiate a `QuantumFloat` with a complex number.

File: src/qrisp/qtypes/quantum_float.py [hunk 6]
```diff
@@ -455,201 +553,259 @@ def sb_poly(self, m=0):
...
 
-        super().encode(encoding_number, permit_dirtyness=permit_dirtyness)
+            Unlike the base :meth:`QuantumVariable.encode
+            <qrisp.QuantumVariable.encode>`, this method additionally accepts
+            ``rounding`` (inserted before ``permit_dirtyness``, for
+            historical reasons specific to QuantumFloat).
 
-    @gate_wrap(permeability="args", is_qfree=True)
-    def __mul__(self, other):
+            Unlike the base's ``ScalarLike``, this does not accept ``complex``
```
Comment (renezander90) [inline src/qrisp/qtypes/quantum_float.py]: Same comment as for `encoder`.

**Comment by purva-thakre** [comment]
@PietropaoloFrisoni Does this also fix https://github.com/eclipse-qrisp/Qrisp/issues/842 ?

File: src/qrisp/qtypes/quantum_float.py [hunk 7]
```diff
@@ -455,201 +553,259 @@ def sb_poly(self, m=0):
...
-            else:
-                output_qf = QuantumFloat(
-                    self.msize + int(np.ceil(np.log2(abs(other)))),
-                    self.exponent,
-                    signed=False,
-                )
+            output_qf = QuantumFloat(
+                self.msize + int(np.ceil(np.log2(abs(other)))),
+                self.exponent,
+                signed=bool(self.signed or other < 0),
```
Comment (renezander90) [inline src/qrisp/qtypes/quantum_float.py:645]: This raises an interesting question (for the future): `signed` is static. Hence, Jasp multiplication cannot make this distinction.

File: src/qrisp/qtypes/quantum_float.py [hunk 8]
```diff
@@ -455,201 +553,259 @@ def sb_poly(self, m=0):
...
+            Suppresses the error message when calling encode on dirty
+            qubits. The default is False.
+
+        Returns
+        -------
+        None
+
+        """
+        value = encoding_number
+        if rounding:
```
Comment (renezander90) [inline src/qrisp/qtypes/quantum_float.py:600]: Wouldn't it be a better policy to always round/truncate?
This would also avoid the situation, that we cannot raise an Error in tracing mode, if the value that should be encoded does not fit. (Likely out of scope for this PR). @purva-thakre @positr0nium

File: src/qrisp/qtypes/quantum_float.py [hunk 9]
```diff
@@ -975,51 +1171,55 @@ def significant(self, k):
...
+            mode (depending on whether the exponent is non-negative, see
+            :meth:`decoder <qrisp.QuantumFloat.decoder>`), or a traced
+            ``jax.Array`` while tracing.
 
         Examples
         --------
-        We create a QuantumFloat and round a value to fit the encoder and subsequently
-        initiate:
+        We create a QuantumFloat and round a value to the closest one it can
+        represent. Note that directly encoding an unrepresentable value (like
```
Comment (renezander90) [inline src/qrisp/qtypes/quantum_float.py:1196]: Okay, this solves #842 @purva-thakre 

However, I think this "truncate silently" that is currently happening is not always "proper" truncation. This brings us back to the issue that it would be easier to always call `truncate` in the encoder (i.e. the rounding always happens).

File: src/qrisp/qtypes/quantum_float.py (hunk 5 above)
Comment (PietropaoloFrisoni) [reply src/qrisp/qtypes/quantum_float.py]: You are right, in my head it was typed since I have a local branch with some type hints in it (`QuantumVariable` will hopefully be improved soon as we have an onboarding ticket for this)

File: src/qrisp/qtypes/quantum_float.py (hunk 3 above)
Comment (PietropaoloFrisoni) [reply src/qrisp/qtypes/quantum_float.py]: This function is not actually used within this file at all, so it would be weird to mark it internal and only use it somewhere else. I turned it into an internal helper and moved it in the only file that uses it

File: src/qrisp/qtypes/quantum_float.py (hunk 9 above)
Comment (PietropaoloFrisoni) [reply src/qrisp/qtypes/quantum_float.py:1196]: I did not touch the default (non-rounding) path in this PR, so that is issue is not fully solved unfortunately. I only focused on improving the performance and correctness of the explicit `rounding=True` option.

If I understand correctly, closing that issue would mean changing the default behavior to reject non-representable values, which is probably a real compatibility change worth its own PR : )

File: src/qrisp/qtypes/quantum_float.py (hunk 8 above)
Comment (PietropaoloFrisoni) [reply src/qrisp/qtypes/quantum_float.py:600]: I can't help with that as I am not familiar enough with quantum arithmetic (one of the components I still need to investigate properly). At least now I have a basic understanding of how encoding/decoding works and what multiplication, summation etc. with a `QuantumFloat` actually means

**Comment by positr0nium** [comment]
> @renezander90 @purva-thakre  I will not request Copilot's reviews anymore since it's basically a rabbit hole. It keeps finding issues, and most of them are pre-existing edge cases that will never come up in practical usage, or they are out of scope for this PR (which was meant to be a quick refactoring but it already costed too much time and tokens)

https://github.com/user-attachments/assets/8795d639-4808-4529-bf43-38d3686e5686

**Comment by positr0nium** [comment]
Jokes aside, I think we need a deeper refactoring to stop this whac-a-mole

**Comment by PietropaoloFrisoni** [comment]
> Jokes aside, I think we need a deeper refactoring to stop this whac-a-mole

I think we should start from the most primitive objects in qrisp (such as `QuantumVariable`) and refactor them systematically:

- Applying `pylint`/`ruff`/`pyright` improvements as much as possible
- Applying type hints as much as possible
- Add proper test coverage for every piece of source code
- Revisit the entire structure and modularization of the test suite and testing pipeline
- Make sure that edge cases are considered and accounted for 
- (...)

Only the first 2 points made us discover so many bugs in this PR and in previous refactorings: 
- https://github.com/eclipse-qrisp/Qrisp/pull/770
- https://github.com/eclipse-qrisp/Qrisp/pull/671
- https://github.com/eclipse-qrisp/Qrisp/pull/526
etc.

One year ago, this would have been daunting for 3-4 people, but now we will soon be more, and we have AI tools. I think this is definitely an investment that is worth doing! I hope we can coordinate efforts in this direction with Fraunhofer as well @renezander90

File: src/qrisp/qtypes/quantum_float.py [hunk 10]
```diff
@@ -334,12 +308,42 @@ class QuantumFloat(QuantumVariable):
...
 
-    def __init__(self, msize, exponent=0, qs=None, name=None, signed=False):
+    signed: bool
+    exponent: int | Array
+    traced_attributes: list[str]
+    static_attributes: list[str]
+
+    def __init__(
+        self,
+        msize: int | Array,
```
Comment (renezander90) [inline src/qrisp/qtypes/quantum_float.py:330]: The type hints a great! However, they could cause some confusion, and user could attempt to initialise a `QuantumFloat` with a "true" array. Maybe, we should add a note in the documentation that this is only relevant in JASP/tracing mode and refers to 0-d (traced) JAX arrays.

**Review by renezander90 — APPROVED** [review]
LGTM 🎊  (It might not solve all problems, but definitively a significant improvement. Thanks!)

**Comment by purva-thakre** [comment]
@renezander90 @diehoq What are your thoughts on us working on a paper about Qrisp's quantum arithmetic module? I brought this up with @positr0nium @PietropaoloFrisoni @moar55. According to Raphael we have historically prioritized other things because the portions of quantum arithmetic shor's algorithm relies on work as expected; which is technically not a great method to gauge the health of a module. 

b0dl4g

The quantum arithmetic bugs we've been addressing are on an ad-hoc basis. I believe we should use this paper as motivation to resolve them comprehensively rather than continuing to deal with them one at a time. 

The paper could also be a good resource for the users who often stumble on `QuantumFloat` technically being `QuantumFixed`. 

Feel free to tag other folks who have worked on quantum arithmetic in Qrisp. Sorry, I do not know all their usernames.

**Comment by PietropaoloFrisoni** [comment]
@purva-thakre I think it would be the perfect motivation/excuse to deep dive into this topic and to improve it in qrisp from scratch (that is, theoretical approach -> efficient implementation for each quantum type -> proper testing for each case that we plan to support)

**Comment by renezander90** [comment]
@purva-thakre this sounds like a great idea! Yes, we always had a big place in our hearts for Shor's algorithm, and even the modular arithmetic that is outside of what is necessary for Shor has been less reliable (recently addressed by @diehoq ). Basically, this should happen while we are refactoring the arithmetic at the same time. (I.e., remove SBP arithmetic and reimplement the primitives that exclusively rely on that like quantum-classical multiplication, migrate all adders to a common interface and testing, consolidate the modular arithmetic, etc.) The goal should be to have the arithmetic in perfect shape and a paper about that (which at the same time can serve as an overview on quantum arithmetic in general) at some point next year.

**Comment by PietropaoloFrisoni** [comment]
> Basically, this should happen while we are refactoring the arithmetic at the same time. (I.e., remove SBP arithmetic and reimplement the primitives that exclusively rely on that like quantum-classical multiplication, migrate all adders to a common interface and testing, consolidate the modular arithmetic, etc.)

We might also take the change to modularize efficiently the jasp pipeline v.s. eager Python code. Right now there are functions that are natively jittable, functions that are not jittable, function that are jittable with a separate pipeline for certain values of the arguments (and the logic is implemented at the same level with the `check_for_tracing_mode` check), etc. We could potentially revisit this structure and define a more consistent logic, at least for quantum arithmetic

File: src/qrisp/qtypes/quantum_float.py (hunk 10 above)
Comment (PietropaoloFrisoni) [reply src/qrisp/qtypes/quantum_float.py:330]: I added this clarification. If you don't like it please let me know, otherwise I will merge this PR in the evening as you already approved : )

**Comment by purva-thakre** [comment]
> We might also take the change to modularize efficiently the jasp pipeline v.s. eager Python code. Right now there are functions that are natively jittable, functions that are not jittable, function that are jittable with a separate pipeline for certain values of the arguments (and the logic is implemented at the same level with the check_for_tracing_mode check), etc. We could potentially revisit this structure and define a more consistent logic, at least for quantum arithmetic

Yes, this has been my plan as a part of working on #501

**Comment by positr0nium** [comment]
Tagging also @bruno-ah-um
